module Temporalio
  class Error < StandardError
    def self.canceled?: (Exception error) -> bool

    def self._with_backtrace_and_cause: (
      Exception err,
      backtrace: Array[String],
      cause: Exception?
    ) -> Exception

    class WorkflowFailedError < Error
      def initialize: -> void
    end

    class WorkflowContinuedAsNewError < Error
      attr_reader new_run_id: String

      def initialize: (new_run_id: String) -> void
    end

    class WorkflowQueryFailedError < Error
    end

    class WorkflowQueryRejectedError < Error
      attr_reader status: Client::WorkflowExecutionStatus::enum

      def initialize: (status: Client::WorkflowExecutionStatus::enum) -> void
    end

    class WorkflowUpdateFailedError < Error
      def initialize: -> void
    end

    class WorkflowUpdateRPCTimeoutOrCanceledError < Error
      def initialize: -> void
    end

    class RPCError < Error
      attr_reader code: Code::enum

      def initialize: (
        String message,
        code: Code::enum,
        raw_grpc_status: String?
      ) -> void

      def grpc_status: -> untyped

      private def create_grpc_status: -> untyped

      module Code
        OK: 0
        CANCELLED: 1
        UNKNOWN: 2
        INVALID_ARGUMENT: 3
        DEADLINE_EXCEEDED: 4
        NOT_FOUND: 5
        ALREADY_EXISTS: 6
        PERMISSION_DENIED: 7
        RESOURCE_EXHAUSTED: 8
        FAILED_PRECONDITION: 9
        ABORTED: 10
        OUT_OF_RANGE: 11
        UNIMPLEMENTED: 12
        INTERNAL: 13
        UNAVAILABLE: 14
        DATA_LOSS: 15
        UNAUTHENTICATED: 16

        type enum = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16
      end
    end
  end
end